core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -499,9 +500,25 @@ private synchronized void ensureSnapshotsLoaded() {
       List<Snapshot> loadedSnapshots = Lists.newArrayList(snapshotsSupplier.get());
       loadedSnapshots.removeIf(s -> s.sequenceNumber() > lastSequenceNumber);
 
-      // Format version 1 does not have accurate sequence numbering, so remove based on timestamp
+      // Format version 1 does not have accurate sequence numbering, so remove based on
+      // ref reachability
       if (this.formatVersion == 1) {
-        loadedSnapshots.removeIf(s -> s.timestampMillis() > currentSnapshot().timestampMillis());
+        Set<Long> snapshotsToKeep = Sets.newHashSet();
```

Comment:
I think we probably want to use something like the following because it ensures that snapshots known at the time of the original table load are present and retained.

```java
      // Format version 1 does not have accurate sequence numbering, so remove based on timestamp
      if (this.formatVersion == 1) {
        List<Long> knownSnapshots = snapshotLog.stream().map(HistoryEntry::snapshotId)
                .collect(Collectors.toList());
        loadedSnapshots.removeIf(s -> !knownSnapshots.contains(s.snapshotId()));
      }
```

Reply:
This might even work for v2, so it's possible we could just use this instead of relying on sequence number

Reply:
It looks like snapshot log is only tracking the main branch currently.  We might need to discuss adding refs to that.